File: Manipulator3D_C_Sharp/src/sim/Trajectory.cs
using System;
using System.Numerics;

namespace Manipulator3D.sim;

public sealed class LinearTrajectory
{
    private Vector3 _a;
    private Vector3 _b;
    private float _duration = 1.0f;
    private float _t;
    private float _alpha;
    private bool _finished = true;

    public void Reset(Vector3 from, Vector3 to, float durationSec)
    {
        _a = from;
        _b = to;
        _duration = durationSec > 1e-6f ? durationSec : 1e-6f;
        _t = 0.0f;
        _alpha = 0.0f;
        _finished = false;
    }

    public void Update(float dt)
    {
        if (_finished) return;
        _t += dt;
        _alpha = Math.Clamp(_t / _duration, 0.0f, 1.0f);
        if (_alpha >= 1.0f) _finished = true;
    }

    public Vector3 Position()
    {
        return _a + (_b - _a) * _alpha;
    }

    public bool Finished() => _finished;
    public float Alpha() => _alpha;
}

File: Manipulator3D_Scala/src/main/scala/manipulator3d/sim/Trajectory.scala
package manipulator3d.sim

import manipulator3d.math.Vec3
import manipulator3d.math.Vec3.clamp

final class LinearTrajectory {
  private var a: Vec3 = Vec3.Zero
  private var b: Vec3 = Vec3.Zero
  private var duration: Float = 1.0f
  private var t: Float = 0.0f
  private var alpha: Float = 0.0f
  private var finished: Boolean = true

  def reset(from: Vec3, to: Vec3, durationSec: Float): Unit = {
    a = from
    b = to
    duration = if (durationSec > 1e-6f) durationSec else 1e-6f
    t = 0.0f
    alpha = 0.0f
    finished = false
  }

  def update(dt: Float): Unit = {
    if (finished) return
    t += dt
    alpha = clamp(t / duration, 0.0f, 1.0f)
    if (alpha >= 1.0f) finished = true
  }

  def position: Vec3 =
    Vec3(
      a.x + (b.x - a.x) * alpha,
      a.y + (b.y - a.y) * alpha,
      a.z + (b.z - a.z) * alpha
    )

  def isFinished: Boolean = finished
  def getAlpha: Float = alpha
}
